src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStorage.java

```diff
@@ -0,0 +1,169 @@
+package org.jenkinsci.plugins.pipeline.milestone;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.Action;
+import hudson.model.Executor;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Manages persistence of active milestones and cancellation of older builds as needed.
+ */
+public abstract class MilestoneStorage implements ExtensionPoint {
+    private static final Logger LOGGER = Logger.getLogger(MilestoneStorage.class.getName());
+
+    /**
+     * @param buildNumber The build number currently querying for builds to cancel
+     * @param ordinal The ordinal the build just passed. {@code null} means it just started.
+     * @param milestones A map keyed by build numbers recording their current milestone.
+     * @return A subset of build numbers among the given milestones eligible for cancellation.
+     */
+    public static Set<Integer> getBuildsToCancel(int buildNumber, @CheckForNull Integer ordinal, @NonNull Map<Integer, Integer> milestones) {
+        return milestones.entrySet().stream()
+                .filter(entry -> entry.getKey() < buildNumber)
+                .filter(entry -> entry.getValue() == null || (ordinal != null && entry.getValue() < ordinal))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toSet());
+    }
+
+
+    /**
+     * Records passing a milestone.
+     * @param run The run passing the milestone.
+     * @param ordinal the ordinal of the milestone gettting passed. May be {@code null} to record build starting.
+     * @return The current list of milestones for the job.
+     */
+    protected abstract Map<Integer, Integer> store(@NonNull Run<?,?> run, @CheckForNull Integer ordinal);
+
+    /**
+     * Records passing a milestone.
+     * <p>
+     * This is expected to cancel all prior builds that did not pass the same milestone,
```

Comment:
```suggestion
     * This will cancel all prior builds that did not pass the same milestone,
```
(since this is not an SPI method)